articles/realtime.md

```diff
@@ -0,0 +1,293 @@
+published: false
+author: Jackie Kay
+---
+
+* This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+Robotic systems need to be responsive. In mission critical applications, a delay of less than a
+millisecond in the system can cause a catastrophic failure. For ROS 2.0 to capture the needs
+of the robotics community, the core software components must not interfere with the requirements of real-time computing.
+
+# Definition of Real-time Computing
+
+Real-time software guarantees correct computation at the correct time. Hard real-time software systems
+have a set of strict deadlines, and missing a deadline is considered a failure. Soft real-time systems
+degrade their quality of service if a deadline is missed.
+
+Real-time computer systems are often associated with low-latency systems. Many applications of real-time
+computing are also low-latency applications (for example, automated piloting systems must be reactive to
+sudden changes in the environment). However, it is
+generally agreed upon that a real-time system is not defined by low latency, but by a deterministic schedule:
+it must be guaranteed that the system finishes a certain task by a certain time. Therefore, it is
+important that the latency in the system be measurable and a maximum allowable latency for tasks be
+set.
+
+A more useful metric in evaluating the "hardness" of a real-time system is jitter, which is defined
+as the variable deviation between a task's deadline and its actual time of completion. A more jittery
+system is less deterministic, less predictable, and less real-time. Though in practice it is impossible
+to completely eliminate jitter from a real-time system, it is a worthy goal to determine a hard upper bound
+for jitter.
+
+# Implementation of Real-time Computing
+
+In general, an operating system can guarantee that the tasks it handles for the developer, such as
+thread scheduling, are deterministic, but the OS may not guarantee that the developer's code will
+run in real-time. Therefore, it is up to the developer to know what the determinstic guarantees
+of an existing system are, and what she must do to write hard real-time code on top of the OS.
+
+In this section, various strategies for developing on top of a real-time
+OS are explored, since these strategies might be applicable to ROS 2. The patterns focus on the
+use case of C/C++ development on Linux-based real-time OS's (such as RTLinux), but the general
+concepts are applicable to other platforms.
+
+## Memory management
+
+### Lock memory, prefault stack:
+
+```c
+    if (mlockall(MCL_CURRENT|MCL_FUTURE) == -1) {
+            perror("mlockall failed");
+            exit(-2);
+    }
+    unsigned char dummy[MAX_SAFE_STACK];
+
+    memset(dummy, 0, MAX_SAFE_STACK);
+``` `
```

Comment:
yep. the syntax highlighting was screwed up in my editor, so I stupidly added another backtick. maybe I should switch to sublime.  7d77b05